constraint/pkg/client/crds/validate.go

```diff
@@ -84,5 +86,16 @@ func ValidateCR(cr *unstructured.Unstructured, crd *apiextensions.CustomResource
 		return fmt.Errorf("%w: %v", constraints.ErrSchema, err.ToAggregate())
 	}
 
+	// validate that there are no unknown fields in the CR
+	// NewStructural assumes the schema is validate
+	structural, err := schema.NewStructural(crd.Spec.Validation.OpenAPIV3Schema)
+	if err != nil {
+		return fmt.Errorf("%w: %v", constraints.ErrInvalidConstraint, err)
+	}
+	unknownFields := structuralpruning.PruneWithOptions(cr.DeepCopy().Object, structural, true, schema.UnknownFieldPathOptions{TrackUnknownFieldPaths: true})
+	if len(unknownFields) > 0 {
+		return fmt.Errorf("%w: %v", constraints.ErrInvalidConstraint, fmt.Sprintf("unknown fields: %v", unknownFields))
```

Comment:
wouldn't this have backward compatibility issues?

Reply:
Great question! The k8s api server already does this kind of validation. One can run into this error by trying to apply a constraint with some random field via `kubectl` today. However, `gator` lacks that behavior. For context, this work is being done as part of https://github.com/open-policy-agent/gatekeeper/issues/2863

So, in short, it is not backwards incompatible wrt to the `on cluster` behavior. Wrt to the `off cluster` behavior, mmmaybe but it's not respecting the openapi v3 schema.

@maxsmythe wdy?

Reply:
To nit, the API server doesn't reject unknown fields, or raise an error, it silently prunes them. Kubectl raises errors.

I don't think this would be backwards compatibility so much as surfacing undetected bugs in users' code. If users are using legacy schema for parameters, then the behavior is unaltered (since we explicitly allow for unknown fields). If users have unknown fields elsewhere:

* The field may be pruned by the API server
* Kubectl may already be giving them warnings
* They may have spelling errors that would otherwise have gone un-noticed (`enfrocementAction`)

Because we control the schema and semantic meaning of constraint fields (other than parameters, which we've already covered), and because the API server already prunes unknown fields, I don't think previous uses of unrecognized fields could be considered valid inputs subject to backwards compatibility protection.

Reply:
thanks for chiming in Max and for making the clear distinctions.

@ritazh does that answer your concern?

Reply:
Yes, thanks for the explanation.